File: ruby/union_find/quick_find.rb
class QuickFind
  attr_reader :parents, :length

  def initialize(n)
    @parents = Array.new(n) { |i| i }
    @length = n
  end

  def connected?(a, b)
    parents[a] == parents[b]
  end

  def union(a, b)
    a_root = parents[a]
    b_root = parents[b]
    return if a_root == b_root

    parents.each_with_index do |x, i|
      parents[i] = a_root if x == b_root
    end

    @length -= 1
  end
end

# uf = QuickFind.new(10)
# uf.union(4, 3)
# uf.union(3, 8)
# uf.union(6, 5)
# uf.union(9, 4)
# uf.union(2, 1)

# fail unless uf.length == 5
# fail unless uf.connected?(0, 7) == false
# fail unless uf.connected?(8, 9) == true

# uf.union(5, 0)
# uf.union(7, 2)
# uf.union(6, 1)
# uf.union(1, 0)

# fail unless uf.length == 2
# fail unless uf.connected?(0, 7) == true

File: rust/union_find/quick_find.rs
pub struct QuickFind {
  parents: Vec<usize>,
  length:  usize,
}

impl QuickFind {
  pub fn new(n: usize) -> QuickFind {
    Self {
      parents: (0..n).collect::<Vec<_>>(),
      length:  n,
    }
  }

  pub fn is_connected(&self, a: usize, b: usize) -> bool {
    return self.parents[a] == self.parents[b];
  }

  pub fn union(&mut self, a: usize, b: usize) {
    let a_root = self.parents[a];
    let b_root = self.parents[b];

    if a_root == b_root {
      return;
    }

    for i in 0..self.parents.len() {
      if self.parents[i] == b_root {
        self.parents[i] = a_root;
      }
    }

    self.length -= 1;
  }
}

// pub fn main() {
//   let mut uf = QuickFind::new(10);
//   uf.union(4, 3);
//   uf.union(3, 8);
//   uf.union(6, 5);
//   uf.union(9, 4);
//   uf.union(2, 1);

//   assert!(uf.length == 5);
//   assert!(uf.is_connected(0, 7) == false);
//   assert!(uf.is_connected(8, 9) == true);

//   uf.union(5, 0);
//   uf.union(7, 2);
//   uf.union(6, 1);
//   uf.union(1, 0);

//   assert!(uf.length == 2);
//   assert!(uf.is_connected(0, 7) == true);
// }